src/Workspaces/Remote/Core/RemoteCallback.cs

```diff
@@ -87,23 +87,75 @@ public async ValueTask<TResult> InvokeAsync<TResult>(Func<T, CancellationToken,
         }
 
         /// <summary>
-        /// Invokes API on the callback object hosted in the original process (usually devenv) associated with the currently executing brokered service hosted in ServiceHub process.
-        /// The API streams results back to the caller.
+        /// Invokes API on the callback object hosted in the original process (usually devenv) associated with the
+        /// currently executing brokered service hosted in ServiceHub process. The API streams results back to the
+        /// caller.
         /// </summary>
-        /// <inheritdoc cref="BrokeredServiceConnection{TService}.InvokeStreamingServiceAsync"/>
+        /// <param name="invocation">A callback to asynchronously write data. The callback should not complete the <see
+        /// cref="PipeWriter"/>, but no harm will happen if it does.</param>
+        /// <param name="reader">A callback to asynchronously read data. The callback should not complete the <see
+        /// cref="PipeReader"/>, but no harm will happen if it does.</param>
+        /// <param name="cancellationToken">A cancellation token the operation will observe.</param>
         public async ValueTask<TResult> InvokeAsync<TResult>(
             Func<T, PipeWriter, CancellationToken, ValueTask> invocation,
             Func<PipeReader, CancellationToken, ValueTask<TResult>> reader,
             CancellationToken cancellationToken)
         {
             try
             {
-                return await BrokeredServiceConnection<T>.InvokeStreamingServiceAsync(_callback, invocation, reader, cancellationToken).ConfigureAwait(false);
+                return await InvokeWorkerAsync(_callback).ConfigureAwait(false);
             }
             catch (Exception exception) when (ReportUnexpectedException(exception, cancellationToken))
             {
                 throw OnUnexpectedException(exception, cancellationToken);
             }
+
+            async ValueTask<TResult> InvokeWorkerAsync(T service)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var pipe = new Pipe();
+
+                // Kick off the work to do the writing to the pipe in a fire-and-forget fashion.  It will start hot and
+                // will be able to do work as the reading side attempts to pull in the data it is writing.
+
+                _ = WriteAsync(service, pipe.Writer);
```

Comment:
this is the least pleasant part of the PR, but talking to @AArnott we think this is correct/fine.  Effectively, the previous code had a serial dependency on writing/reading which we don't want.  We simply want to allow the writing to kick off, allowing it to work in parallel with the reading that happens below.  the reader/writer can communicate with each other, ensuring that if one faults the other will as well.    as both are async, there's no concern of deadlock here.  the reader yields when it has no more data, and hte writer will yield as necessary when flushing if waiting on the reader.

Reply:
@AArnott i just realized a small issue here.  we're kicking off this writing synchronously.  say it starts writing to the pipe and the pipe gets 'full' (not sure the right terminology).  Since we haven't kicked off reading yet, will this just deadlock ehre?  Or will the pipe *always* allow synchronous writing without problem up through the first FlushAsync that we eventually call?

I presume the FlushAsync will then make everything ok as it may yield the writer and then allow the code here to proceed to the point that it starts reading.  then, once enough is read the writer's flush will finish and it will go to writing the next thing.  At that point, we'll be on threadpool threads and should be able to run concurrently.

If the above is an issue, it would be trivial to yield WriteAsync immediately to ensure it's always concurrent with the reading code.

Reply:
> Or will the pipe always allow synchronous writing without problem up through the first FlushAsync that we eventually call?

The PipeWriter will pause when you call `FlushAsync` until the reader starts reading. If your `WriteAsync` method never yields, writes a bunch and then calls `FlushAsync` in a blocking way, such that the reader can never start, yes that would deadlock.

Reply:
wdym by `and then calls FlushAsync in a blocking way,`?

Reply:
By that I mean it blocks a thread instead of awaiting.

Reply:
ok.  in this case, i think we're fine.  all calls to FlushAsync happen in a standard non-blocking way using normal async/await.